core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -768,84 +782,113 @@ private synchronized SplitAssignment getSplitAssignment(PlanNodeId planNodeId, i
+                if (!getTaskStatus().getState().isTerminatingOrDone()) {
+                    scheduleAsyncCleanupRequest("abort", true);
+                }
             }
+        }
+    }
+
+    @Override
+    public void cancel()
+    {
+        // Only cancel the task if we aren't already attempting to abort or fail the task remotely
+        if (!terminating.compareAndSet(false, true)) {
+            return;
+        }
 
-            // send cancel to task and ignore response
-            scheduleAsyncCleanupRequest(new Backoff(maxErrorDuration), "cancel", false);
+        synchronized (this) {
+            try (SetThreadName ignored = new SetThreadName("HttpRemoteTask-%s", taskId)) {
+                TaskStatus taskStatus = getTaskStatus();
+                if (!taskStatus.getState().isTerminatingOrDone()) {
+                    scheduleAsyncCleanupRequest("cancel", false);
+                }
+            }
         }
     }
 
-    private void cleanUpTask()
+    private void cleanUpTask(TaskState taskState)
     {
-        checkState(getTaskStatus().getState().isDone(), "attempt to clean up a task that is not done yet");
+        checkState(taskState.isTerminatingOrDone(), "attempt to clean up a task that is not terminating or done: %s", taskState);
 
         // clear pending splits to free memory
         synchronized (this) {
             pendingSplits.clear();
             pendingSourceSplitCount = 0;
             pendingSourceSplitsWeight = 0;
-            partitionedSplitCountTracker.setPartitionedSplits(PartitionedSplitsInfo.forZeroSplits());
+            partitionedSplitCountTracker.setPartitionedSplits(getPartitionedSplitsInfo());
             splitQueueHasSpace = true;
             whenSplitQueueHasSpace.complete(null, executor);
         }
 
         // clear pending outbound dynamic filters to free memory
         outboundDynamicFiltersCollector.acknowledge(Long.MAX_VALUE);
 
-        // cancel pending request
-        Future<?> request = currentRequest.getAndSet(null);
-        if (request != null) {
-            request.cancel(true);
-        }
-
-        taskStatusFetcher.stop();
-
-        // The remote task is likely to get a delete from the PageBufferClient first.
-        // We send an additional delete anyway to get the final TaskInfo
-        scheduleAsyncCleanupRequest(new Backoff(maxErrorDuration), "cleanup", true);
-    }
+        // only when termination is complete do we shut down status fetching
+        if (taskState.isDone()) {
+            // stop continuously fetching task status
+            taskStatusFetcher.stop();
+            // cancel pending request
+            Future<?> request = currentRequest.getAndSet(null);
+            if (request != null) {
+                request.cancel(true);
+            }
 
-    @Override
-    public synchronized void abort()
-    {
-        if (getTaskStatus().getState().isDone()) {
-            return;
+            taskStatusFetcher.stop();
```

Comment:
Why doing it twice? Please add a comment if it is not a mistake

Reply:
Nope, that's a mistake- will fix.